Split `--tunnel` flag to allow setting a tunneling protocol for a subset of traffic (#24561)
cilium-team/cilium: Documentation/operations/upgrade.rst

  applications create new connections after the TTL specified by the upstream
  DNS server is expired.

New Options
~~~~~~~~~~~

* ``routing-mode=native``: This option enables native-routing mode, in place of
  ``tunnel=disabled``, now deprecated.
* ``tunnel-protocol``: This option allows setting the tunneling protocol, in place
  of e.g., ``tunnel=vxlan``.

Deprecated Options
~~~~~~~~~~~~~~~~~~

> * The ``tunnel`` option was deprecated and will be removed in v1.14. To enable

Reviewer: Did you mean v1.15?

Author: Ouch, yes. That's because this is a rebase of a PR I already attempted to send into v1.13.

> * The ``tunnel`` option was deprecated and will be removed in v1.14. To enable

Reviewer 2: I think it would help to specify the version in which the option was deprecated. For example:

```suggestion
* The ``tunnel`` option was deprecated in v1.1<x> and will be removed in v1.14. To enable
```

Author: I'm not sure I understand this one, but maybe the confusion comes from the typo here: this sentence should say "will be removed in v1.15", as per Chris's comment above.

Then, this whole section of the guide is about v1.14. Do we really want to specify that each item in the list is happening in v1.14?

Author: Hm, maybe the tense is the issue here. I keep using past tense when the other entries are using present or present continuous. Maybe that would help clarify we're still talking about v1.14?

Reviewer 2: > maybe the confusion comes from the typo here: this sentence should say "will be removed in v1.15", as per Chris's comment above.

There's no confusion about the version in which the option will be _removed_. I saw Chris's feedback before I reviewed. My suggestion was specifically to indicate the version in which the option was _deprecated_ (presumably v1.14).

> Then, this whole section of the guide is about v1.14. Do we really want to specify that each item in the list is happening in v1.14?

If the entire context of the document is explicitly specific to v1.14 and unlikely to be reused in a different, non-specific context, then I think it's fine to ignore my suggestion.

However: looking at the file, I see nothing to indicate a specific release in the title or description. Rather, I see that the page is implied to be version-specific because of the branch in which the file will reside when merged. 

The challenge, then, is that when this page is updated in 1.15 and subsequent versions, there will be nothing to indicate the version in which the options were removed--because the version in which deprecation occurred is implied, not explicit.

Does that make sense?

> I keep using past tense when the other entries are using present or present continuous. 

It's not a problem of verb tense agreement; it's a problem of implied vs. explicit. 

> Maybe that would help clarify we're still talking about v1.14?

I certainly think so. 😄

Author: This document isn't all specific to v1.14, but the section where those changes are in is for v1.14 only. See line 312 where the title is (this is a lot clearer on the rendered docs than on GitHub of course).

Once we branch for v1.15, a new v1.15 section will be added here. These entries will not be reused for other versions. This PR is also on purpose not marked for backport so it won't appear in older stable branches.

I'm not against being more specific on versions in sentences if we still think it's needed, but that doesn't look specific to this change. I'd prefer if we address it for all entries in a separate PR if we go that direction.

I'll fix up the bad verb tense.

Reviewer 3: (FWIW I agree with @zacharysarah here that we should just be more explicit, but yeah seems fine to just update them all at once in another PR to set that expectation moving forwards)

  native-routing mode, you can set ``routing-mode=native`` (previously

>   ``tunnel=disabled``). To configure the tunneling protocol, you can set

Reviewer 2: Great example of "to accomplish X, do Y" writing. A small tweak will tighten it up.
```suggestion
  ``tunnel=disabled``). To configure the tunneling protocol, set
```

>   ``tunnel=disabled``). To configure the tunneling protocol, you can set
>   ``tunnel-protocol=geneve`` (previously ``tunnel=geneve``).

Reviewer 4: Should also mention `routing-mode=tunnel` here?
```suggestion
  ``tunnel=disabled``). To configure the tunneling protocol, you can set
  ``routing-mode=tunnel`` and ``tunnel-protocol=geneve`` (previously ``tunnel=geneve``).
```

Author: `routing-mode=tunnel` is already the default value (I said something different in a commit description; it was one of Julian's comments).

* The ``securityContext`` for Hubble Relay now defaults to drop all
  capabilities and run as non-root user.
* The ``containerRuntime.integration`` value is being deprecated in favor of ``bpf.autoMount.enabled``.

> * Following the deprecation of the ``tunnel`` agent flag, ``tunnel`` as been

Reviewer 2: Again, I think it would help to specify the version in which the flag was deprecated. One typo fix:
```suggestion
* Following the deprecation of the ``tunnel`` agent flag in v1.1<x>, ``tunnel`` has been
```

> * Following the deprecation of the ``tunnel`` agent flag, ``tunnel`` as been
>   deprecated in favor of ``routingMode`` and ``tunnelProtocol`` and will be
>   removed in v1.14.

Reviewer 5: ```suggestion
* Following the deprecation of the ``tunnel`` agent flag, ``tunnel`` has been
  deprecated in favor of ``routingMode`` and ``tunnelProtocol``, and will be
  removed in v1.14.
```

Reviewer 2: +1 on the typo fix, but no comma needed.
